### PR TITLE
Cache the asset graph after each build, and read it in at startup

### PR DESCRIPTION
**Pull request by jakemac53**
closes https://github.com/dart-lang/build/issues/7

### CODE REVIEW

File: lib/src/generate/build_impl.dart [hunk 1]
```diff
@@ -79,27 +83,45 @@ class BuildImpl {
...
 
-  /// Asset containing previous build outputs.
-  AssetId get _buildOutputsId =>
-      new AssetId(_packageGraph.root.name, '.build/build_outputs.json');
+  /// Asset containing previous asset dependency graph.
+  AssetId get _assetGraphId =>
+      new AssetId(_packageGraph.root.name, '.build/asset_graph.json');
+
+  /// Reads in the [assetGraph] from disk.
+  Future _readAssetGraph() async {
```
Comment (kegluneq) [inline lib/src/generate/build_impl.dart]: `Future<AssetGraph>`

File: lib/src/generate/build_impl.dart [hunk 2]
```diff
@@ -79,27 +83,45 @@ class BuildImpl {
...
+      /// TODO(jakemac): Only invalidate nodes which need invalidating, which
+      /// will give us incremental builds on startup.
+      var generatedNodes =
+          graph.allNodes.where((node) => node is GeneratedAssetNode);
+      for (var node in generatedNodes) {
+        node.needsUpdate = true;
+      }
+      return graph;
+    } on AssetGraphVersionException catch (_) {
+      return new AssetGraph();
```
Comment (kegluneq) [inline lib/src/generate/build_impl.dart]: In context, this is a reasonable return, but taken alone it has a bad code smell. Add a comment explaining why ignoring the error & returning an empty `AssetGraph` is the right way to handle this.

Is it worthwhile to log the error?

File: lib/src/generate/build_impl.dart [hunk 3]
```diff
@@ -79,27 +83,45 @@ class BuildImpl {
...
+  /// Reads in the [assetGraph] from disk.
+  Future _readAssetGraph() async {
+    if (!await _reader.hasInput(_assetGraphId)) return new AssetGraph();
+    try {
+      _logger.info('Reading cached asset graph.');
+      var graph = new AssetGraph.deserialize(
+          JSON.decode(await _reader.readAsString(_assetGraphId)));
+      /// TODO(jakemac): Only invalidate nodes which need invalidating, which
+      /// will give us incremental builds on startup.
+      var generatedNodes =
```
Comment (kegluneq) [inline lib/src/generate/build_impl.dart]: Feels a bit strange to mix `where` & `for (var ... in ...)`

Maybe:
```dart
graph.allNodes
    .where((node) => node is GeneratedAssetNode)
    .forEach((node) {
      node.needsUpdate = true;
    });
```

File: lib/src/generate/watch_impl.dart [hunk 4]
```diff
@@ -126,40 +127,53 @@ class WatchImpl {
...
+          AssetId id, AssetId primaryInput, ChangeType rootChangeType) async {
         var node = _assetGraph.get(id);
         if (node == null) return;
-        if (node is GeneratedAssetNode) {
-          node.needsUpdate = true;
-        }
         _assetCache.remove(id);
         for (var output in node.outputs) {
-          clearNodeAndDeps(output, rootChangeType);
+          await clearNodeAndDeps(output, primaryInput, rootChangeType);
```
Comment (kegluneq) [inline lib/src/generate/watch_impl.dart]: At the risk of premature optimization, all of these can be processed concurrently.

```dart
await Future.wait(node.outputs.map((output) => clearNodeAndDeps(output, primaryInput, rootChangeType));
```

File: lib/src/generate/watch_impl.dart [hunk 5]
```diff
@@ -126,40 +127,53 @@ class WatchImpl {
...
       }
-      for (var input in updatedInputs.keys) {
-        clearNodeAndDeps(input, updatedInputs[input]);
-      }
-      updatedInputs.clear();
 
-      _logger.info('Starting build');
-      _currentBuild = _buildImpl.runBuild();
-      _currentBuild.then((result) {
+      () async {
```
Comment (kegluneq) [inline lib/src/generate/watch_impl.dart]: This feels a bit strange in Dart, I think the preferred pattern here is `new Future(() { ... });` but I'm not 100% sure.

Comment (jakemac53) [inline lib/src/generate/watch_impl.dart]: Basically, I just wanted to be able to use async/await for this logic, but I need the first part of this function to run synchronously so I didn't want to add async to the entire thing. ie: my end goal was not to schedule all this work for later (although this does have that effect), but instead just be able to use async/await.

File: lib/src/generate/watch_impl.dart (hunk 4 above)
Comment (jakemac53) [inline lib/src/generate/watch_impl.dart]: I thought there might be some race conditions here, but now that I think harder on it I don't think there will be, so I can update this.

File: lib/src/generate/build_impl.dart (hunk 1 above)
Comment (jakemac53) [inline lib/src/generate/build_impl.dart]: done

File: lib/src/generate/build_impl.dart (hunk 3 above)
Comment (jakemac53) [inline lib/src/generate/build_impl.dart]: done

File: lib/src/generate/build_impl.dart (hunk 2 above)
Comment (jakemac53) [inline lib/src/generate/build_impl.dart]: done, logged a message as well

**Comment by jakemac53** [comment]
updated

File: lib/src/generate/watch_impl.dart (hunk 5 above)
Comment (kegluneq) [inline lib/src/generate/watch_impl.dart]: This still looks odd to me. If you prefer to leave the IIFE, please add a comment explaining why.

**Comment by kegluneq** [comment]
LGTM pending added comment re: IIFE

File: lib/src/generate/watch_impl.dart (hunk 5 above)
Comment (jakemac53) [inline lib/src/generate/watch_impl.dart]: I just changed it, also updated to use `Future.wait` which makes the use of async/await less necessary.

**Pull request merged**